group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+            for (Uuid topicId: allSubscribedTopics) {
+                Set<Integer> assignedPartitions = targetAssignment.get(member).targetPartitions().get(topicId);
+                for (int i = 0; i < subscribedTopicDescriber.numPartitions(topicId); i++) {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, i);
+                    if (assignedPartitions == null || !assignedPartitions.contains(i)) {
+                        String otherMember = partitionOwnerInTargetAssignment.get(topicIdPartition);
+                        int otherMemberPartitionCount = assignmentManager.targetAssignmentSize(otherMember);
+                        if (memberPartitionCount + 1 < otherMemberPartitionCount) {
+                            LOG.debug("{} can be moved from member {} to member {} for a more balanced assignment.",
+                                topicIdPartition, otherMember, member);
+                            return false;
+                        }
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Balance the current assignment using the data structures created in the assignPartitions(...) method above.
+     */
+    private void balance() {
+        if (!unassignedPartitions.isEmpty()) unassignedPartitionsAssignment();
+        // Refill unassigned partitions will all the topicId partitions.
+        unassignedPartitions.addAll(topicIdPartitions(subscriptionIds, subscribedTopicDescriber));
+
+        // Narrow down the reassignment scope to only those partitions that can actually be reassigned.
+        Set<TopicIdPartition> fixedPartitions = new HashSet<>();
+        for (Uuid topicId: subscriptionIds) {
+            if (!canTopicParticipateInReassignment(topicId)) {
+                for (int i = 0; i < subscribedTopicDescriber.numPartitions(topicId); i++) {
+                    fixedPartitions.add(new TopicIdPartition(topicId, i));
+                }
+            }
+        }
+        unassignedPartitions.removeAll(fixedPartitions);
+
+        // Narrow down the reassignment scope to only those members that are subject to reassignment.
+        for (String member: members.keySet()) {
+            if (!canMemberParticipateInReassignment(member)) {
+                sortedMembersByAssignmentSize.remove(member);
+            }
+        }
+
+        // If all the partitions are fixed i.e. unassigned partitions is empty there is no point of re-balancing.
+        if (!unassignedPartitions.isEmpty() && !isBalanced()) performReassignments();
+    }
+
+    private void performReassignments() {
+        boolean modified;
+        boolean reassignmentOccurred;
+        // Repeat reassignment until no partition can be moved to improve the balance.
+        do {
+            // Before re-starting the round of reassignments check if the assignment is already balanced.
+            if (isBalanced()) break;
```

Comment:
When `performReassignments` is called for the first time, we check this. However, we also just checked it as part at L400 as well. Should we just remove it there?

Reply:
I see where you are coming from,  I was actually hoping to avoid the method call and local vars initialization altogether if it's balanced, not sure how much it really helps in terms of optimization, what do you think? I also thought it would make logical sense immediately for anyone reading

Reply:
I had addressed removing it from the check before calling perform reassignments again, added to the condition now, thanks for pointing it out

Reply:
I actually just removed the check, it' not necessary